app/components/build/AvatarWithUnknownEmailPrompt.js

```diff
@@ -0,0 +1,337 @@
+  };
+
+  getUserEmail(email) {
+    const userEmails = this.props.viewer.emails.edges;
+
+    const userEmail = userEmails.find(
+      ({ node: { address: userEmail } }) => (
+        userEmail.toLowerCase() === email.toLowerCase()
+      )
+    );
+
+    if (userEmail) {
+      return userEmail.node;
+    }
+  }
+
+  isCurrentUsersValidatedEmail(email) {
+    const userEmail = this.getUserEmail(email);
+    return userEmail && userEmail.verified;
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return shallowCompare(this, nextProps, nextState);
+  }
+
+  componentDidMount() {
+    const { createdBy } = this.props.build;
+
+    if (createdBy && createdBy.email && !this.isCurrentUsersValidatedEmail(createdBy.email)) {
+      this.props.relay.setVariables(
+        {
+          isTryingToPrompt: true,
+          emailForPrompt: createdBy.email.toLowerCase()
+        },
+        (readyState) => {
+          if (readyState.done) {
+            setTimeout(() => {
+              const { dismissedAt } = this.props.viewer.notice;
+
+              if (this._dropdown && !dismissedAt) {
+                this._dropdown.setShowing(true);
+              }
+            }, 0);
+          }
+        }
+      );
+    }
+  }
+
+  handleMutationFailure = (transaction) => {
+    this.setState({ isAddingEmail: false, isDismissing: false, isSendingVerification: false });
+
+    FlashesStore.flash(FlashesStore.ERROR, transaction.getError());
+  };
+
+  handleDismissClick = () => {
+    this.setState({ isDismissing: true });
+
+    const mutation = new NoticeDismissMutation({ viewer: this.props.viewer, notice: this.props.viewer.notice });
+
+    Relay.Store.commitUpdate(mutation, { onSuccess: this.handleDismissedSucess, onFailure: this.handleMutationFailure });
+
+    this._dropdown.setShowing(false);
+  };
+
+  handleAddEmailClick = () => {
+    this.setState({ isAddingEmail: true });
+
+    const mutation = new EmailCreateMutation({ address: this.props.build.createdBy.email, viewer: this.props.viewer });
+
+    Relay.Store.commitUpdate(mutation, { onSuccess: this.handleEmailAddedSuccess, onFailure: this.handleMutationFailure });
+  };
+
+  handleResendVerificationClick = () => {
+    this.setState({ isSendingVerification: true });
+
+    const mutation = new EmailResendVerificationMutation({ email: this.getUserEmail(this.props.build.createdBy.email) });
+
+    Relay.Store.commitUpdate(mutation, { onSuccess: this.handleVerificationResentSuccess, onFailure: this.handleMutationFailure });
+  };
+
+  handleDismissedSucess = () => {
+    this.setState({ isDismissing: false });
+  };
+
+  handleEmailAddedSuccess = () => {
+    this.setState({ isAddingEmail: false, hasSentSomething: true });
+  };
+
+  handleVerificationResentSuccess = () => {
+    this.setState({ isSendingVerification: false, hasSentSomething: true });
+  };
+
+  handleLocalDismissClick = () => {
+    this.setState({ hasBeenDismissed: true });
+  }
+
+  getMessages(loading) {
+    const {
```

Comment:
I actually find declaring all the things like this at the top much harder to read than just calling `this.props.blah` in the function.

Take this line for example:

```
// If we haven't decided to send a query for this yet, don't show anything!
if (!isTryingToPrompt) {
  return {};
}
```

As you scroll your eyes down the function, your constantly having to look back up to the top of the function to see where the value actually came from. Where as:

```
// If we haven't decided to send a query for this yet, don't show anything!
if (!this.state.isTryingToPrompt) {
  return {};
}
```

You brain is instantly: "ah cool, so we're checking state here. rad"